File: scripts/adapter/sidecar/compactor/compactor.py
#!/usr/bin/env python3
"""
Remora Memory Compactor V2.2 (Modular Split Version)
"""
import json
import time
import sqlite3
import os
import sys
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "..")))

import argparse

from schema.schema_init import DB_PATH, DATA_DIR, init_db
from adapter.maintenance.session_gc import prune_expired_watermarks
from adapter.maintenance.topic_gc import run_garbage_collection

from sidecar_lock import acquire_lock, release_lock
from extract_decisions import process_sessions, AgentApiError
from sync_artifacts import scan_and_ingest_artifacts
from check_approval import check_plan_approval
from consume_events import consume_event_queue
from core.storage.topics import get_all_project_uuids

def prune_sidecar_events():
    """清理 Antigravity 系统分发给 Sidecar 的过剩僵尸审计日志"""
    try:
        events_dir = os.path.join(DATA_DIR, "events")
        if os.path.exists(events_dir):
            import glob
            for f in glob.glob(os.path.join(events_dir, "*.json")):
                try:
                    os.remove(f)
                except Exception:
                    pass
    except Exception:
        pass

def main():
    parser = argparse.ArgumentParser(description="Remora Memory Compactor V2.2")
    parser.add_argument("--cron", action="store_true")
    parser.add_argument("--event-driven", action="store_true")
    args = parser.parse_args()

    init_db()

    if args.event_driven:
        # Stop 钩子同步扫描制品。它是单次极轻量的，无需文件锁保护，避免互斥冲突
        try:
            context = json.load(sys.stdin)
            scan_and_ingest_artifacts(context)
        except Exception:
            pass
    else:
        # 默认或 --cron 阶段的后台增量对话扫描，长耗时，必须文件锁保护
        acquire_lock()
        cycle_start = time.time()
        try:
            prune_expired_watermarks()
            process_sessions(cycle_start)
            
            # [P0] 串行保序：前置 decisions 提取 commit 之后，立即执行 Plan 审批拦截与事件队列 AI 精确匹配
            with sqlite3.connect(DB_PATH, timeout=15) as conn:
                active_projects = get_all_project_uuids(conn)
                for p_uuid in active_projects:
                    check_plan_approval(conn, p_uuid)
                consume_event_queue(conn, cycle_start)
                run_garbage_collection(conn)
        except AgentApiError as e:
            print(str(e), file=sys.stderr)
            release_lock()
            sys.exit(1)
        except Exception:
            import traceback
            traceback.print_exc()
        finally:
            prune_sidecar_events()
            release_lock()

if __name__ == "__main__":
    main()

File: packages/adapter-antigravity/src/sidecar/compactor.ts
/**
 * Remora Memory Compactor V2.2 (Modular Split Version)
 */

import * as fs from "node:fs";
import * as path from "node:path";
import { getDataDir, getAntigravityDir } from "../bridge/paths";
import { initDb } from "../schema/schema-init";

import { pruneExpiredWatermarks } from "../maintenance/session-gc";
import { runGarbageCollection } from "../maintenance/topic-gc";

import { acquireLock, releaseLock } from "./sidecar-lock";
import { processSessions, AgentApiError } from "./extract-decisions";
import { scanAndIngestArtifacts } from "./sync-artifacts";
import { checkPlanApproval } from "./check-approval";
import { consumeEventQueue } from "./consume-events";
import { getConn, getAllProjectUuids } from "@remora/core";

export function pruneSidecarEvents(): void {
	try {
		const pluginName = process.env.ANTIGRAVITY_PLUGIN_NAME ?? "remora-plugin";
		const eventsDir = path.join(
			getAntigravityDir(),
			"sidecar_data",
			pluginName,
			"memory-compactor",
			"events",
		);
		if (fs.existsSync(eventsDir)) {
			for (const f of fs.readdirSync(eventsDir)) {
				if (f.endsWith(".json")) {
					try {
						fs.unlinkSync(path.join(eventsDir, f));
					} catch (e) {
						console.error("[Remora Daemon Error] Exception in loop:", e);
						continue;
					}
				}
			}
		}
	} catch (e) {
		console.error("[Remora Daemon Error] Exception initializing sidecar events pruning:", e);
	}
}

function main(): void {
	const args = process.argv.slice(2);
	const eventDriven = args.includes("--event-driven");

	initDb();

	if (eventDriven) {
		try {
			const stdin = fs.readFileSync(process.stdin.fd, "utf-8");
			const context = JSON.parse(stdin);
			scanAndIngestArtifacts(context);
		} catch (e) {
			console.error("[Remora Daemon Error] Exception initializing event driven mode:", e);
		}
	} else {
		acquireLock();
		const cycleStart = Date.now() / 1000;
		try {
			pruneExpiredWatermarks();
			processSessions(cycleStart);

			const conn = getConn();
			try {
				const activeProjects = getAllProjectUuids(conn);
				for (const pUuid of activeProjects) {
					checkPlanApproval(pUuid, conn);
				}
				consumeEventQueue(cycleStart, conn);
				runGarbageCollection(conn);
			} finally {
				conn.close();
			}
		} catch (e) {
			if (e instanceof AgentApiError) {
				process.stderr.write(String(e) + "\n");
				releaseLock();
				process.exit(1);
			}
			console.error(e);
		} finally {
			pruneSidecarEvents();
			releaseLock();
		}
	}
}

export { main };

if (typeof require !== "undefined" && require.main === module) {
	main();
}
